### PR TITLE
SC-143 login app adds session tags

### PR DESCRIPTION
**Pull request by brucehoff**
This should be merged only after https://github.com/Sage-Bionetworks/synapse-login-aws-infra/pull/32

This PR makes the login app' add session tags from the OIDC user claims returned by Synapse upon login.  Note that we change from using 'AssumeRoleWithWebIdentity' to simply 'AssumeRole', since the latter allows us to add tags.

### CODE REVIEW

**Comment by zaro0508** [comment]
I'm a bit confused about your comment in the commit message, "Note that we change from using 'AssumeRoleWithWebIdentity' to simply 'AssumeRole', since the latter allows us to add tags".  
Before I start reviewing this PR could you please clarify what you mean by that?  I read the [AWS session tags doc](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_session-tags.html#id_session-tags_operations) and there is a section "Passing Session Tags using AssumeRoleWithWebIdentity".  From reading that page it seems like we should actually be using `AssumeRoleWithWebIdentity` instead of just `AssumeRole`?

**Comment by brucehoff** [comment]
> I'm a bit confused about your comment in the commit message, "Note that we change from using 'AssumeRoleWithWebIdentity' to simply 'AssumeRole', since the latter allows us to add tags".
> Before I start reviewing this PR could you please clarify what you mean by that? I read the [AWS session tags doc](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_session-tags.html#id_session-tags_operations) and there is a section "Passing Session Tags using AssumeRoleWithWebIdentity". From reading that page it seems like we should actually be using `AssumeRoleWithWebIdentity` instead of just `AssumeRole`?

That is a good question:  To pass tags using AssumeRoleWithWebIdentity the tags have to be baked into the user identity token that the login app' passes to AWS.  See "Example Decoded JSON Web Token" [here](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_session-tags.html#id_session-tags_adding-assume-role-idp).  As you can see, there needs to be a special claim called "https://aws.amazon.com/tags", which obviously would not be in the identity token returned by Synapse.     So the login app would have to create such a token, sign it and pass it to AWS.  AWS then needs to validate the token, which means it has to link the "issuer" of the token to a web server address which publishes the public key, retrieve the public key, and validate the signature in the token.  To implement this we would have to basically set up another OpenID Connect provider, securely store the signing key, publish the validation key, etc.  It would work but it is a lot of machinery to add just to pass session tags.  The obvious question is, "why have two different ways to pass session tags then?"  The answer is that the AssumeRoleWithWebIdentity feature is meant to support *public* (not confidential) OAuth clients, i.e. clients that cannot store an OAuth client secret or AWS credentials (think of an installed mobile device app').  In our case the login app' *is* a confidential client so we can use the simpler approach of using the AssumeRole service.

File: src/main/resources/scipooldev.properties [hunk 1]
```diff
@@ -2,7 +2,8 @@ SYNAPSE_OAUTH_CLIENT_ID=100055
 SYNAPSE_OAUTH_CLIENT_SECRET=ssm::/synapse-login-scipooldev/SynapseOauthClientSecret
 AWS_REGION=us-east-1
 SESSION_TIMEOUT_SECONDS=28800
-USER_CLAIMS=userid
+SESSION_NAME_CLAIMS=userid
+SESSION_TAG_CLAIMS=sub,userid,user_name
```
Comment (zaro0508) [inline src/main/resources/scipooldev.properties]: might be nice to add team and company to help with cost evaluations?

File: src/test/java/synapseawsconsolelogin/AuthTest.java [hunk 2]
```diff
@@ -43,15 +43,15 @@
...
+		System.setProperty("SESSION_TAG_CLAIMS", "userid,user_name");
 		System.setProperty(Auth.PROPERTIES_FILENAME_PARAMETER, "test.properties");
 	}
 	
 	@After
 	public void after() {
 		System.clearProperty("TEAM_TO_ROLE_ARN_MAP");
 		System.clearProperty("AWS_REGION");
-		System.clearProperty("USER_CLAIMS");
+		System.clearProperty("SESSION_TAG_CLAIMS");
```
Comment (zaro0508) [inline src/test/java/synapseawsconsolelogin/AuthTest.java:65]: you replaced the `USER_CLAIMS` test with a test for `SESSION_TAG_CLAIMS`.  What about a test for the new  `SESSION_NAME_CLAIMS`?

File: src/main/java/synapseawsconsolelogin/Auth.java [hunk 3]
```diff
@@ -277,38 +305,24 @@ private void doGetIntern(HttpServletRequest req, HttpServletResponse resp)
...
-								}};
-						}
-						@Override
-						public void refresh() {}}).build();
-			
-			AssumeRoleWithWebIdentityResult assumeRoleWithWebIdentityResult = stsClient.assumeRoleWithWebIdentity(assumeRoleWithWebIdentityRequest);
-			Credentials credentials = assumeRoleWithWebIdentityResult.getCredentials();
+					.withRegion(Regions.fromName(awsRegion)).build();
+			Collection<Tag> tags = new ArrayList<Tag>();
+			tags.add(new Tag().withKey("foo").withValue("bar"));
```
Comment (zaro0508) [inline src/main/java/synapseawsconsolelogin/Auth.java]: what's up with this?  the client just needs some dummy tags?

**Review by zaro0508 — CHANGES_REQUESTED** [review]
Thanks for the clarification.  Looks good, just had a few questions.

File: src/main/resources/scipooldev.properties (hunk 1 above)
Comment (brucehoff) [reply src/main/resources/scipooldev.properties]: Done!

Comment (tthyer) [inline src/main/resources/scipooldev.properties]: We may also want `firstName` and `lastName` from userProfile to tag resources provisioned through Service Catalog, mostly as a convenience for IT.

**Review by tthyer — CHANGES_REQUESTED** [review]

File: src/main/resources/scipooldev.properties (hunk 1 above)
Comment (brucehoff) [reply src/main/resources/scipooldev.properties]: Done!

File: src/test/java/synapseawsconsolelogin/AuthTest.java (hunk 2 above)
Comment (brucehoff) [reply src/test/java/synapseawsconsolelogin/AuthTest.java:65]: Good suggestion.  Will be included in the PR update.

File: src/main/java/synapseawsconsolelogin/Auth.java (hunk 3 above)
Comment (brucehoff) [reply src/main/java/synapseawsconsolelogin/Auth.java]: It's a bug.  I factored out the code into a testable method and added the needed test.

**Review by zaro0508 — APPROVED** [review]

**Pull request merged**